src/gallium/auxiliary/util/u_box.h

```diff
@@ -124,34 +124,46 @@ static inline void
 u_box_union_2d(struct pipe_box *dst,
                const struct pipe_box *a, const struct pipe_box *b)
 {
-   int x, y;
-
-   x = MIN2(a->x, b->x);
-   y = MIN2(a->y, b->y);
-
-   dst->width = MAX2(a->x + a->width, b->x + b->width) - x;
-   dst->height = MAX2(a->y + a->height, b->y + b->height) - y;
-   dst->x = x;
-   dst->y = y;
+   if (!a->width || !a->height) {
+      memcpy(dst, b, sizeof(struct pipe_box));
+   } else if (!b->width || !b->height) {
```

Comment:
In the function, you can have dst == a, or dst == b.
memcpy with dst == src seems to be an 'undefined behaviour', which means you have to do another check to avoid the case.
You can also do:
*dst = *b;
I guess it is ok when dst==b ? (not sure 100% sure though)

When the nine state trackers checks for empty box, it justs checks for width == 0. That's because it knows for its use cases checking that is enough (when width is 0, all other fields will be 0 too).

To avoid this issue of aliasing handling, and reduce the number of checks, I tend to think now that the first approach you took was better, ie handle the empty box case in nine. (but applied to all u_box_union_2d usages in nine).